transaction.go

```diff
@@ -0,0 +1,489 @@
+		6 + // timestamp
+		1 + // input count
+		1 + // factoid output count
+		1 // EC output count
+
+	TransactionTotalMinLen = TransactionHeadMinLen // Coinbases have no body
+
+)
+
+// Decode will consume as many bytes as necessary to unmarshal the factoid
+// transaction. It will return the number of bytes read and an error.
+func (f *FactoidTransaction) Decode(data []byte) (i int, err error) {
+	// Because the length of an FactoidTransaction is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("failed to unmarshal")
+		}
+	}()
+
+	if len(data) < TransactionTotalMinLen {
+		return 0, fmt.Errorf("insufficient length")
+	}
+
+	// Decode header
+	i, err = f.DecodeHeader(data)
+	if err != nil {
+		return 0, err
+	}
+
+	// Decode the inputs
+	f.FCTInputs = make([]FactoidTransactionIO, f.InputCount)
+	read, err := FactoidTransactionIOs(f.FCTInputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the FCT Outputs
+	f.FCTOutputs = make([]FactoidTransactionIO, f.FCTOutputCount)
+	read, err = FactoidTransactionIOs(f.FCTOutputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the EC Outputs
+	f.ECOutputs = make([]FactoidTransactionIO, f.ECOutputCount)
+	read, err = FactoidTransactionIOs(f.ECOutputs).Decode(data[i:])
+	if err != nil {
+		return 0, err
+	}
+	i += read
+
+	// Decode the signature blocks, one per input
+	f.Signatures = make([]FactoidTransactionSignature, f.InputCount)
+	for c := uint8(0); c < f.InputCount; c++ {
+		// f.Signatures[i] = new(FactoidTransactionSignature)
+		read, err := f.Signatures[c].Decode(data[i:])
+		if err != nil {
+			return 0, err
+		}
+		i += read
+	}
+
+	txid, err := f.ComputeTransactionID()
+	if err != nil {
+		return 0, err
+	}
+	f.TransactionID = &txid
+
+	return i, err
+}
+
+// UnmarshalBinary unmarshals the data into a factoid transaction.
+func (f *FactoidTransaction) UnmarshalBinary(data []byte) error {
+	// TODO: Some length checks to prevent too few/too many bytes
+	_, err := f.Decode(data)
+	return err
+}
+
+func (f *FactoidTransaction) DecodeHeader(data []byte) (int, error) {
+	version, i := varintf.Decode(data)
+	f.Version = version
+
+	msdata := make([]byte, 8)
+	copy(msdata[2:], data[i:i+6])
+	ms := binary.BigEndian.Uint64(msdata)
+	f.Timestamp = time.Unix(0, int64(ms)*1e6)
+	i += 6
+	f.InputCount = data[i]
+	i += 1
+	f.FCTOutputCount = data[i]
+	i += 1
+	f.ECOutputCount = data[i]
+	i += 1
+
+	return i, nil
+}
+
+// Decode takes a given input and decodes the set of bytes needed to populate
+// the set of factoid transactions ios. The set length should be preset before
+// calling this function. It will return how many bytes it read and return an error.
+func (ios FactoidTransactionIOs) Decode(data []byte) (int, error) {
+	i := 0
+	for c := 0; c < len(ios); c++ {
+		read, err := ios[c].Decode(data[i:])
```

Comment:
I wonder if all this Decode nesting couldn't be reduced. but we'll leave it for now. However it might be a good idea to unexport a lot of these Decode functions to simplify the exported API. Outside of this package I doubt anyone will need to decode just a FactoidTransactionIOs. That slice type may also not need to be exported either. Anything that won't be explicitly needed outside the package should not be exported to keep the API small.

Reply:
Those are good points. I would like to keep the Decode on the tx exported. For myself as a developer, I dislike when a libary thinks I do not ever need something. But the nested decodes we could probably hide, like the slice implementations.

Reply:
Although the API exported is exported on the structures themselves, not the top level. So it's not a lot of overhead on the user to keep it in. But I can live with hiding a lot of these decodes.

Reply:
Any exposed API is something that we later may be locked into. That's why I want to keep it small.

I think it's important to keep in mind the specific use case of these data structures to the users of this package: primarily factom clients, specifically fatd. If it isn't clearly needed for those two, I'd like to keep it unexported. It is easier to add these things later than remove them.

Reply:
I hid the FactoidTransactionsIO data type that I use to not have to write the same for loop each time. Let me know which functions you'd like to hide.